src/net/ip.cc

```diff
@@ -193,6 +193,9 @@ ipv4::handle_received_packet(packet p, ethernet_address from) {
             auto cpu_id = this_shard_id();
             auto l4 = _l4[h.ip_proto];
             if (l4) {
+                if (smp::count == 1) {
```

Comment:
Maybe the `_netif->rss_key().empty()` one?

Reply:
this is not fragile. from a more practical perspective, if we have a single core, there is no point of doing RSS at all. as i think, RSS (receive side scaling) is designed for multi-core settings.

but the proposed change works as well. as this is the only indication if RSS is enabled or not. just a side note, we also support virtio device with the native tcp/ip stack. which does not override the `rss_key()` method, so our virtio device still suffers from this problem even with the proposed `_netif->rss_key().empty()` fix: we are still hashing the packet to a single core.

@techtricity hi John, after reading 
```c++
cpu_id = _netif->hash2cpu(toeplitz_hash(_netif->rss_key(), hash_data));
```

i think even we have only a single core, it still works, as `hash2cpu()` always hashes the number to the `smp::count` cores. what do you mean by "the call to toeplitz_hash faults" exactly? does it mean that we `toeplitz_hash()` for no goods, but it still works or it leads to some unexpected behavior like package loss or wrong order of package delivery?

Reply:
> Making this check be symmetrical with some not-apparently-related one from another file doesn't look correct.
> 

speaking of "symmetrical", probably might be better off mirroring the structure of `dpdk_device::init_port_start()` 
```
if (smp::count > 1) {
  /// do the RSS stuff
} else {
  /// otherwise
}
```
see https://github.com/scylladb/seastar/blob/811505f6b772fa5b620981d4526c1ad2df8ee612/src/net/dpdk.cc#L1515C5-L1515C26 ?

but i don't feel strong either way.

Reply:
> this is not fragile. from a more practical perspective, if we have a single core, there is no point of doing RSS at all. as i think, RSS (receive side scaling) is designed for multi-core settings.

Absolutely agree. However, if we're on multi-core this does _not_ automatically mean RSS is enabled, isn't it? The fact that now it's always enable on smp is just because dpdk.cc's init_port_start() doesn't have another option to turn RSS off

Reply:
It's not possible to turn RSS off with smp > 1. In Linux you can do it (and just lock sockets), but here scokets have no locks.

Reply:
i see. so in other words, the ip stack's implementation should not be hardwired to a certain device's implementation's behavior. it's more related to a design philosophy and probably maintainability.

i am convinced by your point now.

@techtricity hi John, what do you think?

Reply:
So from what @avikivity is saying, RSS is **always** enabled on multi-cpu (good to know), but that means then it **won't** be enabled on single-cpu. 

If that is the case, then on single cpu, the code should skip over RSS forwarding, which is what my fix does. I'm willing to check if RSS enabled and if not just pass packet up, as from @avikivity comment, that is correct, but in a single-cpu you can't forward, so **both methods are correct** (if I'm reading things correctly), it is just what method is preferred - I'm just trying to fix a definite problem that causes code to fault. LMK know what approach the majority wants.

Reply:
On Linux RSS is an optimization/feature. In Seastar it's the main/only data path. So I don't think a separate flag is needed.

Reply:
i am now confused now. if RSS is the main/only data path, do you suggest that we have something like
```c++
inline uint32_t
toeplitz_hash(rss_key_type key, const T& data) {
    // we could use an empty RSS key when smp::count is 1
    if (key.empty()) {
        return 0;
    }
    uint32_t hash = 0, v;
    // do the hash
    return hash;
}
```

Reply:
@tchaikov If Toeplitz can return 0 as a potential hash, then 0 cannot be a return value, as it could collide with the hash value - this was why I was reticent to modify the function. From Wikipedia ( https://en.wikipedia.org/wiki/Toeplitz_Hash_Algorithm ) it sounds like 0 may be a valid hash, but I'm not certain. My thought was to either check if the key was empty **outside** of the data path, or just go with the original method and skip over it all if running single-core... Is the original author of this code around to confirm the best approach?

Reply:
I misled you. RSS can be bypassed if the interface has as many queues as there are shards. So it's possible to have smp > 1 and no RSS.

Reply:
(but, there's no need for a flag to turn RSS off, since there's no alternative)